File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.cpp
#include <bits/stdc++.h>
using namespace std;

int dl, pl; //뭔가 해야하는 마지막집
vector<int> d, p;

void findLast(){
    bool cd = false, cp = false;
    for(int i=dl; i >= 0; i--){
        if(d[i] != 0){
            dl = i;
            cd = true;
            break;
        }
    
    }
    for(int i=pl; i >=0; i--){
        if(p[i] != 0){
            pl = i;
            cp = true;
            break;
        }
    }
    if(!cd) dl = -1;
    if(!cp) pl = -1;
    
}

long long solution(int cap, int n, vector<int> deliveries, vector<int> pickups) {
    
    //뒤에서 부터 배달해주고 뒤에서 부터 빼온다.
    d = deliveries, p = pickups;
    dl = d.size()-1; pl = p.size()-1;
    long long answer = 0;
    
    findLast();
    while(dl != -1 || pl != -1){
        int last = max(dl, pl);
        //가는 동안 뒤에 집부터 배달하기
        int temp = cap;
        for(int i=dl; i >= 0; i--){
            if(d[i] > 0){
                //cap이 남았을 때
                if(temp >= d[i]){
                    temp -= d[i];
                    d[i] = 0;
                }
                //cap이 부족할때
                else{
                    d[i] -= temp;
                    temp = 0;
                    break;
                }
            }
        }
        
        temp = cap;
        for(int i=pl; i >= 0; i--){
            if(p[i] > 0){
                if(temp >= p[i]){
                    temp -= p[i];
                    p[i] = 0;
                }
                else{
                    p[i] -= temp;
                    break;
                }
            }
        }
        
        answer += (last+1) * 2;
        findLast();
    }
    
    
    return answer;
}

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.java
//4시 20분 시작

/*
가면서 제일 끝에다 내리면 됨
돌아올때 그냥 계속 들고오면 됨 -> 근데 수거할게 더 많으면?
deliveries든 pickups 든 가면서 제일 끝집에서 부터 빼면 됨
그게 이동 횟수
*/

class Solution {
    int gcap, gn;
    int dLast, pLast;
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        gcap = cap;
        gn = n;
        
        dLast = getLast(deliveries, n-1);
        pLast = getLast(pickups, n-1);
        
        long answer = 0;
        while (dLast != -1 || pLast != -1) {
            answer +=  (Math.max(dLast, pLast) + 1) * 2;
            update(deliveries, dLast);
            update(pickups, pLast);    
            
            dLast = getLast(deliveries, dLast);
            pLast = getLast(pickups, pLast);
        }
        
        return answer;
    }
    
    public int getLast(int[] arr, int last) {
        for (int i=last; i>=0; i--) {
            if (arr[i] != 0) return i;
        }
        
        return -1;
    }
    
    public void update(int[] arr, int last) {
        if (last == -1) return;
        
        int leftCount = gcap;
        for (int i=last; i>=0; i--) {
            if (arr[i] >= leftCount) {
                arr[i] -= leftCount;
                break;
            } else {
                leftCount -= arr[i];
                arr[i] = 0;
            }
        }
    }
}